### PR TITLE
Make response to failed multiplayer password attempt friendlier

### LINKED ISSUE #14421: Entering wrong password when joining private multi lobby gives error
**video:** https://youtu.be/BZ8h93JxH-g
**osu!lazer version:** 2021.820.0-lazer

### PR DESCRIPTION
**Pull request by sh0ckR6**
Instead of throwing an error, just close the popover and let the user continue with what they were doing
* Resolves #14421

### CODE REVIEW

**Comment by frenzibyte** [comment]
I haven't checked the code in detail and ran it yet, but, from the description, just closing the popover sounds completely bad UX to me (almost feels like bad internet connection preventing client from joining). 

It should at the very least shake the join button (or the popover?) and show a message in red inside the popover that the password entered was incorrect. (cc @peppy)

**Comment by sh0ckR6** [comment]
I 100% agree, looking at this now this feels like really bad UX. I'll update the PR with changes to shake the join button.

I'm also pretty sure I found a bug with the popover while working on this PR, so I've opened a new issue regarding this (#14669)

**Comment by sh0ckR6** [comment]
Updated PR to include popover shaking and error message on attempt fail.

File: osu.Game/Graphics/Containers/ShakeContainer.cs
Comment (bdach) [inline osu.Game/Graphics/Containers/ShakeContainer.cs:33]: Not sure about this event. Maybe `Shake()` should just return a `TransformSequence` to append possible completion callbacks to.

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: This won't work if the user's language on the website is not in English. [The API error message is localised.](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L56)

This should either check the response error code, or web should return a language-invariant error "type" token of some kind.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs:182]: This delegate signature is pretty egregious. Would be good to not have nested `Action`s if at all possible. Maybe the lounge room should resolve the lounge and call `Join()` directly or something of the sort.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs:229]: It's pretty weird to do this in the placeholder text I think? Not sure this is even required, the standard login box just shakes.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Graphics/Containers/ShakeContainer.cs
Comment (sh0ckR6) [reply osu.Game/Graphics/Containers/ShakeContainer.cs:33]: Would it be better to just have the callback as a parameter in `Shake()` rather than returning the `TransformSequence`? Returning the sequence would allow for the caller to add their own transforms to the chain which might be more visibility than intended.

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: [The function that handles authentication for rooms](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L43) that you linked to returns `403` for different reasons (if the user being added to the room is not the currently authenticated user & invalid password). I may have misunderstood what the [first `403` returns](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L47) as I'm not too familiar with PHP, but judging from the naming that looks like what the check is.

Because of the varying `403` responses, having the feedback be the same for an unauthenticated user/invalid user id and an invalid password may be confusing.

A `type` token with the response would work, or a `401 Unauthorized` could be returned instead of a `403` for an invalid password. Either solution works, but would require updates to web before this PR can be merged.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs:229]: Just shaking the login box instead of giving a response in the placeholder text seems like it fits more with the rest of the lazer UX, so I'll take out the placeholder response.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs:182]: I'm thinking I can just pass down the already-resolved `LoungeSubScreen` from `DrawableLoungeRoom` on construction. I'll try it out and see if it works and update the PR if it does.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs]: I've removed the placeholder text response but kept the text clear to make it easier for the user to retry.

File: osu.Game/Graphics/Containers/ShakeContainer.cs
Comment (frenzibyte) [reply osu.Game/Graphics/Containers/ShakeContainer.cs:33]: I disagree, `Shake()` is just like any transform method and should return a `TransformSequence` for appending other transforms to it. (it should even be an extension following the transform patterns in `TransformableExtensions` and `TransformSequenceExtensions` rather than a method, but that can happen separately)

Comment (sh0ckR6) [reply osu.Game/Graphics/Containers/ShakeContainer.cs:33]: Makes sense. I was looking at it like it only belonged in `ShakeContainer`. Considering it _should_ be part of `TransformableExtensions`/`TransformSequenceExtensions`, it makes way more sense to return the `TransformSequence`.

Comment (sh0ckR6) [reply osu.Game/Graphics/Containers/ShakeContainer.cs:33]: Actually, now that the placeholder text response is no longer being used, the event is no longer in use, so I'll remove the event & return `Shake()` to it's original form. The change to return the `TransformSequence` from `Shake()` doesn't make sense here as it doesn't apply anymore, but while we're here I can open an issue/discussion about moving `Shake()` to somewhere in `TransformableExtensions`/`TransformSequenceExtensions` if you'd like.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs:125]: Components aren't usually passed down manually like this. Resolving locally inside the popover (the same way `DrawableLoungeRoom` does) is preferable and should work as well.

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs
Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: 401 is wrong. It means "no credentials provided", and 403 means "credentials provided are wrong".

Also, when I attempted to add test coverage of this, the hard cast to `APIException` will cause crashes if `exception` is not indeed of that type. This should be using a soft cast like the `OperationCanceledException` case does.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: > 401 is wrong. It means "no credentials provided", and 403 means "credentials provided are wrong".

Misread the docs for `401`, my bad!

> Also, when I attempted to add test coverage of this, the hard cast to APIException will cause crashes if exception is not indeed of that type. This should be using a soft cast like the OperationCanceledException case does.

Makes sense, the `if` should be checking if `exception` is actually an `APIException` before trying to cast to it.

---

Also, assuming I'm right about the [first `403` return](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L47), would it actually just be safe to check against a `403` and assume that the `userId` provided by lazer would be the same as the authenticated user? My main hesitation is that I don't want the user to receive the same feedback if their password is wrong (user's fault) vs. lazer is sending the wrong `userId` in the request (internal issue).

Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: Getting error information from the source of the error (i.e. the API/web) is always preferred to implementing heuristics in order to discern what an error means client-side. There's no reason to make changes to web to be able to tell what the error means in a user-language-agnostic manner.

File: osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs]: Updated PR with requested changes.

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs
Comment (sh0ckR6) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: I'm not entirely sure how the `abort()` method works in osu-web, so I'm unsure on how to go about returning a `type` token in the body. If possible, the `DisplayableError` class could have a nullable `JsonProperty` containing the `type` token (as I assume no other endpoint would return this token), which can be checked instead of the message.

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:91]: This implementation also just fails when trying to add a test (which this PR is missing). @sh0ckR6 we require that tests are added to cover new behaviours like this for this precise reason.

**Comment by peppy** [comment]
The shaking of this container is at the wrong level. It causes the contents of the popover to shake rather than the popover itself.

**Comment by peppy** [comment]
Popovers have a very inflexible structure, which means shaking can't really be done right now. I'm going to remove it.

**Comment by peppy** [comment]
@nekodex we don't have a sound effect for "incorrect password", probably need one (buzzer or something).

**Comment by peppy** [comment]
I've added back shake support and made things look/feel a bit better. Needs a re-review from someone other than me.

**Review by peppy — DISMISSED** [review]

**Comment by bdach** [comment]
Amended one xmldoc, looks good to go otherwise. :+1:

**Review by bdach — APPROVED** [review]

**Pull request merged**
